lib/client/signal_fx_client.js

```diff
@@ -0,0 +1,166 @@
+
+  this.queue = [];
+  this.asyncRunning = false;
+}
+
+SignalFxClient.prototype.HEADER_API_TOKEN_KEY = 'X-SF-Token';
+SignalFxClient.prototype.HEADER_USER_AGENT_KEY = 'User-Agent';
+SignalFxClient.prototype.HEADER_CONTENT_TYPE = 'Content-Type';
+SignalFxClient.prototype.INGEST_ENDPOINT_SUFFIX = 'v2/datapoint';
+SignalFxClient.prototype.API_ENDPOINT_SUFFIX = 'v1/event';
+
+/**
+ * Send the given metrics to SignalFx.
+ *
+ * @param data - param object with followinf fields:
+ *    'cumulative_counters' (list): a list of dictionaries representing the cumulative counters to report.
+ *    'gauges' (list): a list of dictionaries representing the gauges to report.
+ *    'counters' (list): a list of dictionaries representing the counters to report.
+ */
+SignalFxClient.prototype.send = function (data) {
+  var _this = this;
+
+  function processDataPoints(metricType, dataPoints) {
+    if (!dataPoints) {
+      return;
+    }
+
+    if (!Array.isArray(dataPoints)) {
+      throw new TypeError('DataPoints not of type list: ' + dataPoints);
+    }
+
+    dataPoints.forEach(function (dp) {
+      _this._addToQueue(metricType, dp);
+    });
+  }
+
+  processDataPoints('cumulative_counter', data.cumulative_counters);
+  processDataPoints('gauge', data.gauges);
+  processDataPoints('counter', data.counters);
+
+
+  this.startAsyncSend();
+};
+
+SignalFxClient.prototype._addToQueue = function (metricType, datapoint) {
+  throw new Error('Subclasses should implement this!');
+};
+
+SignalFxClient.prototype._batchData = function (datapointsList) {
+  throw new Error('Subclasses should implement this!');
+};
+
+/**
+ * Send an event to SignalFx
+ *
+ * @param eventType (string): the event type (name of the event time series).
+ * @param dimensions  (dict): a map of event dimensions.
+ * @param properties  (dict): a map of extra properties on that event.
+ */
+SignalFxClient.prototype.sendEvent = function (eventType, dimensions, properties) {
+  var data = {
+    eventType: eventType,
+    dimensions: dimensions || {},
+    properties: properties || {}
+  };
+
+  var url = this.apiEndpoint + '/' + this.API_ENDPOINT_SUFFIX;
+  this.post(JSON.stringify(data), url, conf.JSON_HEADER_CONTENT_TYPE);
+};
+
+SignalFxClient.prototype.getHeaderContentType = function () {
+  throw new Error('Subclasses should implement this!');
+};
+
+SignalFxClient.prototype.startAsyncSend = function () {
+  if (this.asyncRunning) {
+    return;
+  }
+
+  var _this = this;
+  this.asyncRunning = true;
+
+
+  // Send post request in separate thread
+  async.queue(function (task, callback) {
+    var datapointsList = [];
+    while (_this.queue.length !== 0 && datapointsList.length < _this.batchSize) {
+      datapointsList.push(_this.queue.shift());
+    }
+
+    var dataToSend = _this._batchData(datapointsList);
+    if (dataToSend) {
+      var url = _this.ingestEndpoint + '/' + _this.INGEST_ENDPOINT_SUFFIX;
+      _this.post(dataToSend, url);
+    }
+    callback();
```

Comment:
Shouldn't this callback be called after the post request has completed? Otherwise this is pretty much the same as a for loop (in which case you don't need async.queue).